File: cpp/dp/divide_conquer_dp.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
const ll DCDP_INF = (1LL << 62);
template <class Cost>
void compute(int l, int r, int optL, int optR, const vector<ll> &prev, vector<ll> &cur, Cost cost) {
  if (l > r)
    return;
  int mid = (l + r) / 2;
  pair<ll, int> best = {DCDP_INF, -1};
  for (int k = optL; k <= min(optR, mid); k++)
    best = min(best, {prev[k] + cost(k, mid), k});
  cur[mid] = best.first;
  compute(l, mid - 1, optL, best.second, prev, cur, cost);
  compute(mid + 1, r, best.second, optR, prev, cur, cost);
}

File: python/dp/divide_conquer_dp.py
INF = 10**30


def divide_conquer_dp(prev, n, compute_cost):
    cur = [INF] * n

    def solve(l, r, opt_l, opt_r):
        if l >= r:
            return
        mid = (l + r) // 2
        best = (INF, opt_l)
        for k in range(opt_l, min(opt_r, mid) + 1):
            best = min(best, (prev[k] + compute_cost(k, mid), k))
        cur[mid] = best[0]
        solve(l, mid, opt_l, best[1])
        solve(mid + 1, r, best[1], opt_r)

    solve(0, n, 0, n - 1)
    return cur
